### DISCUSSION TITLE
Recommended parameters for archiving purposes

### DISCUSSION THREAD

**Discussion by odedia** [RhetTbull/osxphotos#1515] [General]
What would be good parameters for archiving purposes?
Basically, in the event that my Photos library is dead, and iCloud is suddenly gone - what would be a good command that would allow my to later do an `osxphotos import` and get back to as close to 100% same results as before?
The library contains RAW images, jpegs, HEICs, live photos, RAW+JPEG combos etc.

This is what I'm currently using for export, I don't have an equivelant import command yet:

```
osxphotos export /Volumes/ExternalData/Users/odedia/Pictures/OSXPhotos \
--directory "{created.year}/{created.year}-{created.mm}-{created.dd}" \
--retry 3 --sidecar XMP \
--keyword-template "{label}" \
--keyword-template "{keyword}" \
--keyword-template "{person}" \
--replace-keywords \
--download-missing --use-photokit \
--edited-suffix “_edited” --update --cleanup \
--exportdb ~/.osxphotos_export.db
```

**Comment by oPromessa** [comment]
I'm doing a bit the same to rebuild my photos library. See #1267

I'm not using label or person on keywords as I'm assuming Photos will recreate them once you import them back. 

I want to recreate the albums so my directory structure includes {album}.

**Comment by RhetTbull** [comment]
As @oPromessa said, if you use albums a lot you might want to capture those.  One way is to use the `--directory {folder_album}` template but this means you lose the date-based organization (and also that a photo in multiple albums gets exported multiple times). Another alternative is to use the album as a keyword. For example, `--keyword-template "{album?album:{folder_album}}"` would create a keyword in form `album:Album Name` and `album:Folder/Subfolder/Album Name` if there's a folder.  Upon re-import to Photos, you could use the `osxphotos batch-edit` command to restore all the albums from the keywords:

`osxphotos batch-edit --verbose --album "{keyword|filter(startswith album:)|sslice(6:)}" --split-folder "/"`

This finds any keyword starting with `album:` then extracts the album (and folders) and recreates the folder structure in Photos.

I think you could do this with the `--album` option of the import command too but would need to test that as the metadata works a bit differently on import (it needs to use metadata or the sidecar to read the metadata as the photo is not yet imported into Photos).

If you use favorites a lot you might create a custom keyword (`--keyword-template "{favorite?Favorite,}"`) or use the `--favorite-rating` option.  (Though there is not yet a corresponding `--favorite-rating` option in `osxphotos import` -- see #1373).

I would also use the `--export-aae` option which exports AAE adjustment files for edited photos. I'm working on a feature now for `osxphotos import` that will automatically import these and cause Photos to re-apply the adjustments to the imported photo (preserving non-destructive edits which the current export command doesn't do when exporting the edited image).

**Comment by odedia** [comment]
I guess that looking forward from an archive perspective - it makes sense to only save the originals (RAW, JPEGs, RAW+JPEG combos) and the .AAE files, but not the edited jpegs (which may come from a cache anyway), so I think a good combo would be `--export-aae --skip-edited`.

**Comment by odedia** [comment]
Seems like exporting AAEs can clash if you have multiple versions of the same image (two virtual copies for example):
```
Error exporting photo (9DA859AF-063F-4307-B105-9439BA87BDD5: _DSC0353.NEF) as _DSC0353.NEF: Error copying file /Volumes/ExternalData/Users/odedia/Pictures/Photo Library.photoslibrary/resources/renders/9/9DA859AF-063F-4307-B105-9439BA87BDD5.plist to /Volumes/ExternalData/Users/odedia/Pictures/OSXPhotos/Originals/2019/2019-10-15/_DSC0353.AAE: Error Domain=NSCocoaErrorDomain Code=516 "“9DA859AF-063F-4307-B105-9439BA87BDD5.plist” couldn’t be copied to “2019-10-15” because an item with the same name already exists." UserInfo={NSSourceFilePathErrorKey=/Volumes/ExternalData/Users/odedia/Pictures/Photo Library.photoslibrary/resources/renders/9/9DA859AF-063F-4307-B105-9439BA87BDD5.plist, NSUserStringVariant=(
    Copy
), NSDestinationFilePath=/Volumes/ExternalData/Users/odedia/Pictures/OSXPhotos/Originals/2019/2019-10-15/_DSC0353.AAE, NSFilePath=/Volumes/ExternalData/Users/odedia/Pictures/Photo Library.photoslibrary/resources/renders/9/9DA859AF-063F-4307-B105-9439BA87BDD5.plist, NSUnderlyingError=0x600003ee0270 {Error Domain=NSPOSIXErrorDomain Code=17 "File exists"}} (photoexporter.pyc: 1227)
```

EDIT: Actually, on the second run it appears that this happens for every single AAE file, it seems like osxphotos cannot identify that it's updating the existing AAE?
I temporarily disabled export_aae in the toml file.

**Comment by hyfen** [comment]
I've been struggling with a similar problem. I'm leaning towards not saving labels into the keywords metadata field since these are not user-generated and may change as Apple's image classification algorithms evolve.
